### PR TITLE
[DAG] SelectionDAG.computeKnownBits - add NSW/NUW flags support to ISD::SHL handling

### LINKED ISSUE #89414: [DAG] SelectionDAG.computeKnownBits - add NSW/NUW flags support to ISD::SHL handling
The call to KnownBits::shl is currently not handling nodes that have the NSW/NUW flags (or a known non zero shift amount).

https://github.com/llvm/llvm-project/blob/e553ac4d8148291914526f4f66f09e362ce0a63f/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp#L3497-L3506

This task will mainly involve adding test coverage - my advice would be to disable the equivalent handling in ValueTracking, run `check-llvm-transforms` to see what tests start failing and see what equivalent tests can be created for CodeGen/X86

https://github.com/llvm/llvm-project/blob/e553ac4d8148291914526f4f66f09e362ce0a63f/llvm/lib/Analysis/ValueTracking.cpp#L1168-L1182

### PR DESCRIPTION
**Pull request by zxc12523**
fix #89414

### CODE REVIEW

**Review by arsenm — COMMENTED** [review]
Needs tests

File: llvm/test/CodeGen/X86/fold-int-pow2-with-fmul-or-fdiv.ll [hunk 1]
```diff
@@ -1668,4 +1668,4 @@ define float @fdiv_pow_shl_cnt32_okay(i32 %cnt) nounwind {
   %conv = uitofp i32 %shl to float
   %mul = fdiv float 0x3a20000000000000, %conv
   ret float %mul
-}
+}
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/fold-int-pow2-with-fmul-or-fdiv.ll:1671]: end of file newline error

File: llvm/test/CodeGen/X86/pr89877.ll [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+; X64-LABEL: sext_known_nonzero:
+; X64:       # %bb.0:
+; X64-NEXT:    movl %edi, %ecx
+; X64-NEXT:    movl $256, %eax # imm = 0x100
+; X64-NEXT:    # kill: def $cl killed $cl killed $ecx
+; X64-NEXT:    shll %cl, %eax
+; X64-NEXT:    movzwl	%ax, %eax
+; X64-NEXT:    rep bsfl %eax, %eax
+; X64-NEXT:    retq
+  %x = shl nuw nsw i16 256, %xx
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/pr89877.ll:28]: Tests should stress the individual nuw/nsw flags. Also should have some negative tests where the required flag is missing

File: llvm/test/CodeGen/X86/pr89877.ll [hunk 3]
```diff
@@ -0,0 +1,126 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 3
+; RUN: llc < %s -mtriple=i686-unknown-unknown -mattr=+sse2 | FileCheck %s --check-prefixes=X86
+; RUN: llc < %s -mtriple=x86_64-unknown-unknown -mattr=+avx | FileCheck %s --check-prefixes=X64
+; RUN: llc < %s -mtriple=x86_64-unknown-unknown | FileCheck %s --check-prefixes=CHECK-SSE
+; RUN: llc < %s -mtriple=x86_64-unknown-unknown -mattr=+avx2 | FileCheck %s --check-prefixes=CHECK-AVX2
+; RUN: llc < %s -mtriple=x86_64-unknown-unknown -mattr=+avx512f | FileCheck %s --check-prefixes=CHECK-NO-FASTFMA
+; RUN: llc < %s -mtriple=x86_64-unknown-unknown -mcpu=skx -fp-contract=fast | FileCheck %s --check-prefixes=CHECK-FMA
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/pr89877.ll:7]: -fp-contract flag shouldn't be necessary for the purposes of the test (if it were, you would be better of applying the contract flag to the individual instructions)

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 4]
```diff
@@ -3492,16 +3492,23 @@ KnownBits SelectionDAG::computeKnownBits(SDValue Op, const APInt &DemandedElts,
...
     break;
   }
-  case ISD::SHL:
+  case ISD::SHL: {
     Known = computeKnownBits(Op.getOperand(0), DemandedElts, Depth + 1);
     Known2 = computeKnownBits(Op.getOperand(1), DemandedElts, Depth + 1);
-    Known = KnownBits::shl(Known, Known2);
+
+    bool NUW = Op.getNode()->Flags.hasNoUnsignedWrap();
+    bool NSW = Op.getNode()->Flags.hasNoSignedWrap();
```
[the remark this answers is not in the document]
Comment (zxc12523) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:3500]: It seems like there is no getFlags in SDValue. Maybe we can add this method?

File: llvm/test/CodeGen/X86/pr89877.ll (hunk 3 above)
Comment (zxc12523) [reply llvm/test/CodeGen/X86/pr89877.ll:7]: When I run "ninja check-llvm-codegen-x86", I found some tasks fail due to the modification. One of those is fold-int-pow2-with-fmul-or-fdiv.ll, which contain "-fp-contract=fast" in the testing command. So, I guess it is not related to this.

File: llvm/test/CodeGen/X86/pr89877.ll (hunk 2 above)
Comment (zxc12523) [reply llvm/test/CodeGen/X86/pr89877.ll:28]: Can u give me an example of negative test?

File: llvm/test/CodeGen/X86/fold-int-pow2-with-fmul-or-fdiv.ll [hunk 5]
```diff
@@ -3,6 +3,7 @@
 ; RUN: llc < %s -mtriple=x86_64-unknown-unknown -mattr=+avx2 | FileCheck %s --check-prefixes=CHECK-AVX,CHECK-AVX2
 ; RUN: llc < %s -mtriple=x86_64-unknown-unknown -mattr=+avx512f | FileCheck %s --check-prefixes=CHECK-AVX,CHECK-AVX512F,CHECK-NO-FASTFMA
 ; RUN: llc < %s -mtriple=x86_64-unknown-unknown -mcpu=skx -fp-contract=fast | FileCheck %s --check-prefixes=CHECK-AVX,CHECK-AVX512F,CHECK-FMA
+; XFAIL: *
```
Comment (RKSimon) [inline llvm/test/CodeGen/X86/fold-int-pow2-with-fmul-or-fdiv.ll:6]: remove this

File: llvm/test/CodeGen/X86/known-never-zero.ll [hunk 6]
```diff
@@ -1,6 +1,7 @@
 ; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
 ; RUN: llc < %s -mtriple=i686-unknown-unknown -mattr=+sse2 | FileCheck %s --check-prefixes=X86
 ; RUN: llc < %s -mtriple=x86_64-unknown-unknown -mattr=+avx | FileCheck %s --check-prefixes=X64
+; XFAIL: *
```
Comment (RKSimon) [inline llvm/test/CodeGen/X86/known-never-zero.ll:4]: remove this

File: llvm/test/CodeGen/X86/pr89877.ll [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+; X86-NEXT:    shll %cl, %eax
+; X86-NEXT:    cwtl
+; X86-NEXT:    testl	%eax, %eax
+; X86-NEXT:    je	.LBB0_1
+; X86-NEXT:  # %bb.2:                                # %cond.false
+; X86-NEXT:	   rep		bsfl	%eax, %eax
+; X86-NEXT:	   retl
+; X86-NEXT:	 .LBB0_1:
+; X86-NEXT:	   movl	$32, %eax
+; X86-NEXT:	   retl
```
Comment (RKSimon) [inline llvm/test/CodeGen/X86/pr89877.ll:19]: this indentation looks off - was this generated by the update_llc_test_checks script?

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (zxc12523) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:3500]: fixed

File: llvm/test/CodeGen/X86/fold-int-pow2-with-fmul-or-fdiv.ll (hunk 1 above)
Comment (zxc12523) [reply llvm/test/CodeGen/X86/fold-int-pow2-with-fmul-or-fdiv.ll:1672]: fixed

File: llvm/test/CodeGen/X86/fold-int-pow2-with-fmul-or-fdiv.ll (hunk 5 above)
Comment (zxc12523) [reply llvm/test/CodeGen/X86/fold-int-pow2-with-fmul-or-fdiv.ll:6]: fixed

File: llvm/test/CodeGen/X86/known-never-zero.ll (hunk 6 above)
Comment (zxc12523) [reply llvm/test/CodeGen/X86/known-never-zero.ll:4]: fixed

File: llvm/test/CodeGen/X86/pr89877.ll (hunk 2 above)
Comment (arsenm) [reply llvm/test/CodeGen/X86/pr89877.ll]: The same test, just with flags removed that shows your patch made no change

Comment (zxc12523) [reply llvm/test/CodeGen/X86/pr89877.ll]: Will I also need to add test case for ShAmtNonZero flag?

**Review by RKSimon — CHANGES_REQUESTED** [review]
According to the CI you need to regenerate fold-int-pow2-with-fmul-or-fdiv.ll and known-never-zero.ll

File: llvm/test/CodeGen/X86/fold-int-pow2-with-fmul-or-fdiv.ll (hunk 1 above)
Comment (RKSimon) [reply llvm/test/CodeGen/X86/fold-int-pow2-with-fmul-or-fdiv.ll:1638]: Still needs an extra empty line at the EOF

**Review by RKSimon — APPROVED** [review]
LGTM

**Pull request merged**